### ISSUE TITLE
Feature_selection: feature shuffling

### ISSUE THREAD

**Issue by solegalli** [solegalli/feature_engine#136]
A transformer with an implementation of this:
https://github.com/solegalli/feature-selection-for-machine-learning/blob/master/11-Hybrid-methods/11.1-Feature-shuffling.ipynb

See what is available in scikit-learn first. I think they have something to derive importance by shuffling.

**Comment by solegalli** [comment]
@KishManani this one?

It is not the most complicated of the selection transformers :p.

Please check out the develop branch and make the PR to develop as well? there is a folder: selection, with some transformers already, check DropDuplicatedFeatures(), which kind of shows which are the default checks to include in init, fit and transform.

Thank you!

**Comment by SunnyxBd** [comment]
@solegalli 

I have created a PR for this issue https://github.com/solegalli/feature_engine/pull/163/commits/8dbde1490ba1da742c6efd93d2c0b6cca1224a14

TO ADD:

- more comments on my code.

- I have put the skeleton the the unit tests file, I didn't write the tests yet.

I need you to review my code to see if we are aligned before I start the unit tests.

Thank you.

**Comment by solegalli** [comment]
Hey @SunnyxBd  thanks for the PR. 

A few things to consider. This transformer's aim is to select features. Not only to shuffle them. They way it selects features is as follows:

1. trains a classifier or regressor using all features in the dataset and determines a specific metric, example roc auc, or rmse
2. takes 1 feature and shuffles it. With the dataframe that now contains a shuffled feature, using the model trained in 1, it determines a new prediction, a new metric (roc, rmse) and compares it with the one from 1
3. determine the performance drift = original value (from step 1) - value with shuffled feature (from step 2)
3. if the decrease in performance of the metric (perf drift from step 3) is above a threshold, then keep feature, otherwise remove
4. repeat for the following feature, 1 at the time, until all features are examined.

(Cells 6 onwards from the notebook attached)

example, I have a dataset with 10 variables
1) train a random forest with all variables and obtain the roc-auc = 0.9
2) shuffle variable 1, use random forest from step 1 to obtain a prediction, and new roc is 0.88
3) calculate performance drift = 0.9 - 0.88 = 0.02
4) if drift bigger than threshold entered by user I keep feature. If threshold is 0.01, then that feature needs to be kept.

in the init, we need to include the following parameters:

- [ ]  variables: the variables the user wants to examine, defaults to None, which means examine all variables

- [ ] classifier or regressor, defaults to random forest

- [ ]  scoring: metric the user wants to evaluate to determine if feature needs to be kept or removed 

- [ ] threshold: the value that defines if a feature will be kept or removed.

- [ ] regression: true or false, if it is a regression or classification dataset (not sure we need this, we need to think)

- [ ] cv: do we want to offer crossvalidation?

(see init of this transformer as example https://github.com/solegalli/feature_engine/blob/master/feature_engine/discretisation/decision_tree.py#L11)

regarding the threshold, there are a few things to consider: 

for roc, auc, r2, precision, recall, etc differences will be between 0 and 1. For mse, mae, they will be big. So I think, it would be useful to add a check there in the init, to control that the threshodl is betwern 0 and 1 for those metrics, or it can be whatever formse and mae (need to have a look at the available metrics in sklearn.metrics)

Another important thing to consider is that a performance drift that is negative for mse means feature was important, but of it is positive for roc means that the feature was important, so we need to be carefull when we select features, of the metric)

in the fit method, we should store the following attributes:

selected_features: the selected features

and maybe a dictionary of features:performance_drift pairs so that the user knows what happened there. It is easier if they need to redefine the threshold.

Thoughts on this?

Thanks a lot for the contribution!

Cheers
Sole

**Comment by SunnyxBd** [comment]
@solegalli Thank you for responding. Yes, this is much clearer.

**Comment by SunnyxBd** [comment]
@solegalli 

I have a question about the estimator.

Do we want the user to pick any estimator for this feature selection method? In other words, do we want to have wherever "sklearn estimator object" as input for init? ( In this case, do we accept estimators that are not from sklearn (from Statsmodels
 for example ?)

Or should it only be between random trees regressors and random tree classifiers?

**Comment by SunnyxBd** [comment]
Also, do we want the parameter _scoring_ to be a sklearn object or a str value?

If it is a sklearn object, we need to add tests to verify it is appropriate for the ML problem (regression or classification)

If it is a str value, we need to pre-define a list of metrics the user can choose from. I'm thinking [auc, f1_score, accuracy, precision, recall, mse, mae, r2_score]. Do you have others in mind?

We also need to verify that the metric is appropriate for the ML problem.

What are your thoughts about this? 
@solegalli

**Comment by solegalli** [comment]
Hi @SunnyxBd I am thinking that if we use sklearn cross_validate, this should handle metrics, estimator and cv  for us:
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html

metrics should be any of the sklearn perfmitted ones:
https://scikit-learn.org/stable/modules/classes.html#classification-metrics

estimators any of the sklearn accepted ones. 

Cross_validate takes as inputs scoring, estimator and cv and would check all, so we shouldn't have to.

The only thing is to set return_estimator=True. This will return the train estimator for each fold. So every time we shuffle a feature, we would have to make predictions with each and average the score? or explore a bit what is the output and see how best to incorporate it.

Would you be happy to try?

Cheers

Sole

**Comment by SunnyxBd** [comment]
@solegalli Ok for sklearn Cross_validate. I'm going to use it.

For the cross validation with feature shuffling, I'll start with averaging scores to get something working. Then, I'll focus on exploring other methods of calculating the "shuffled" score for each feature.

Thank you, 
Sana

**Comment by SunnyxBd** [comment]
@solegalli 

I have added two unit tests for regression.
I have also merged your PR, thank you for your contribution.

Concerning the PR, since we are using the parameter _scoring_ from sklean's _cross_validate_ method, we need to conform to the scoring str mentioned [here](https://scikit-learn.org/stable/modules/model_evaluation.html#scoring)

I believe using the scoring from the list, generalizes model evaluation as follows:
a higher score means a better model performance.

In fact, **mean_absolute_error is replaced by neg_mean_absolute_error** also **mean_squared_error is replaced by ‘neg_mean_squared_error’** etc.

Should I revert your change about calculating the performance drift depending on a specific list of scores?

Can you also check the unit tests I've added and give me your opinion?

Much appreciated,
Sana

**Comment by solegalli** [comment]
@SunnyxBd thanks for the heads-up on the sklearn metrics. I wasn't aware that they negate. That was a very good catch.

I made a few changes and a PR to your repo. I think that would finish this transformer. Let me know otherwise or if there is anything else I missed.

Thanks a lot!

**Comment by SunnyxBd** [comment]
@solegalli 

Did I merge your PR?

I thought I did, but I'm not able to find your changes on my branch,.
I'm also not able to find the PR again.
Can you please verify?
